### DISCUSSION TITLE
forms container and ability to invalidate inputs

### DISCUSSION THREAD

**Discussion by patrickwasp** [zauberzeug/nicegui#801] [Ideas / Feature Requests]
When submitting forms sometimes you don't want to check the validity of each field until the entire form is summitted. Sometimes you can't know if a field is invalid until checking the values with the backend, and you may not want to do that after setting each value individually as it can result in false invalidations. 

something like this maybe

```python
text = ui.input(label='Text', placeholder='start typing',
         on_change=lambda e: result.set_text('you typed: ' + e.value),
         validation={'Input too long': lambda value: len(value) < 20})

ui.button('Click me!', on_click=submit)

def submit():
    valid = text.validate()
    if valid:
        save()
```

A form container would also be helpful to check the validity of all inputs at once

```python
form = ui.form()

with form:
    text = ui.input(label='Text', placeholder='start typing',
            on_change=lambda e: result.set_text('you typed: ' + e.value),
            validation={'Input too long': lambda value: len(value) < 20})
    ui.button('Click me!', on_click=submit)

def submit():
    valid = form.validate()
    if valid:
        save()

```

**Comment by falkoschindler** [comment]
Thanks for your suggestion, @patrickwasp!

It looks like we could use a bit more control over when the validation happens. There has been a related discussion on Discord: https://discord.com/channels/1089836369431498784/1093296231821881364/1094332882996179014
Being able to set the "validation event" to `None` or `[]` would allow us to disable the automatic validation. Then we'd need a `validate()` method to trigger it programmatically.

Introducing a `ui.form` element might be a way to support validation of multiple elements. But do we really need a new element for it? Alternatively we could let the `validate()` method of a normal element traverse its children (if there are any).

**Reply by patrickwasp** [reply]
For some reason, I can't see the discord conversation, when I follow the link I get a message about there not being any text channels. 

Setting the validation event to None sounds like a good idea when you want more control over when validation should happen.

I only bring up a form element as that's the convention I've seen in other frameworks. 
For example, Quasar specifically calls out validation as the purpose for their form component. https://quasar.dev/vue-components/form
> QForm component renders a form DOM element and allows you to easily validate child form components

Form elements also add syntactic sugar for actions like submitting or resetting all related inputs without having to do it explicitly.

**Comment by rodja** [comment]
As mentioned in https://github.com/zauberzeug/nicegui/discussions/1003#discussioncomment-6128707, it could be implemented with a `@ui.on_change` decorator.

**Comment by rodja** [comment]
There is a new feature request about simplifying the gathering of data within a form: https://github.com/zauberzeug/nicegui/discussions/1242. Maybe it is related to validation?

**Comment by rodja** [comment]
@natankeddem has posted two new approaches for form validation in https://github.com/zauberzeug/nicegui/discussions/1791.

**Comment by Ch41r05** [comment]
Hello everyone,

Jumping in on the matter, I'm also facing some issues on input fields validation in forms.
My specific case is inherent to use the validate() method when the on blur event occurs, but this could be extended to also this case scenario, that's why my posting here.

From digging a bit into nicegui internals I found

nicegui\elements\mixins\validation_element.py:48-50

```py
    def _handle_value_change(self, value: Any) -> None:
        super()._handle_value_change(value)
        self.validate()
```

So, from my undestanding, the validation and the on_change properties have a "silent" coupling which I don't know if it's a good way to link the two.

Would removing the self.valiate() from the method (hence using the parent's method) and letting the .on() event listeners manage validation be a bad idea?

Could something like

```py
ui.input(validation={"no value!": lambda e: bool(e)}).on("blur", lambda e: e.sender.validate())
```

be a starting point to make validation triggers decoupled from other events?

**Reply by falkoschindler** [reply]
Hi @patrickwasp,
Since this discussion is almost a year old and focusses on a slightly different topic, I suggest to open a new feature request for this matter. And maybe you can explain a bit more why you think decoupling the validation would be a good thing. Thanks!
